DS-4515: Start new submission via Search or ID Lookup (i.e. submit external source)

sourcedump opened the pull request:
## References
* Link to [JIRA DS-4515](https://jira.lyrasis.org/browse/DS-4515)
* Fixes #720 
* Depends on #708
* REST API: https://github.com/DSpace/DSpace/pull/2824

## Description
Pull request to merge the item submission from external source into the master branch. This feature includes the ability to search and submit an item whose metadata comes from an external source such as Pubmed.

## Instructions for Reviewers
List of changes in this PR:
* A new button in MyDspace page ('Import from external source') that starts the import process. It leads to the page at '/import-external';
* A new page at '/import-external' where it's possible to submit a search to an external source;
* A new modal to preview the external source item metadata and run the import process;
* The most important part of the code is contained in 'src/app/+import-external-page' and 'src/app/submission/import-external'.

NB: from the -collection choice modal- onwards, the features are part of the [Pull 708](https://github.com/DSpace/dspace-angular/pull/708).

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_
- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [X] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [X] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [X] My PR passes all specs/tests and includes new/updated specs for any bug fixes, improvements or new features. A few reminders about what constitutes good tests:
    * Include tests for different user types (if behavior differs), including: (1) Anonymous user, (2) Logged in user (non-admin), and (3) Administrator.
    * Include tests for error scenarios, e.g. when errors/warnings should appear (or buttons should be disabled).
    * For bug fixes, include a test that reproduces the bug and proves it is fixed. For clarity, it may be useful to provide the test in a separate commit from the bug fix.
- [X] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/master/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

tdonohue commented:
@sourcedump and/or @atarix83 : Can we provide additional testing instructions for this PR? It looks like our [PR template](https://github.com/DSpace/dspace-angular/blob/master/.github/pull_request_template.md) was not followed in the description above.  Does this PR require any changes to the REST API, or is this able to run on latest master?

sourcedump commented:
@tdonohue, sorry, my bad, for some reason I've missed the 'template' for the pull request. I'll write the instruction asap.

tdonohue commented:
@sourcedump : Please rebase this on the latest `master`. I've merged #708 (which this depended on), so there's some new merge conflicts above.

sourcedump commented:
Hello @tdonohue! rebase done. I think that we are now ready for the reviews.

sourcedump commented on src/app/core/data/external-source.service.ts:
```diff
@@ -59,6 +60,34 @@ export class ExternalSourceService extends DataService<ExternalSource> {
+   */
+  getAllExternalSources(options: FindListOptions = {}, ...linksToFollow: Array<FollowLinkConfig<ExternalSource>>): Observable<RemoteData<PaginatedList<ExternalSource>>> {
+    return this.findAll(options, ...linksToFollow);
+  }
```
The wrappers around findById and findAll are not necessary, I've removed them.

sourcedump commented on src/app/submission/import-external/import-external-preview/submission-import-external-preview.component.spec.ts:
```diff
@@ -0,0 +1,154 @@
+      spyOn(comp, 'closeMetadataModal');
+      submissionServiceStub.createSubmissionFromExternalSource.and.returnValue(observableOf(submissionObjects));
+      spyOn(compAsAny.router, 'navigateByUrl');
+      comp.import();
```
Ok, thanks, I hope I've solved the problem with the scheduler.

sourcedump commented:
Hello!
I understand your concern but the two buttons should exist separately because the two processes are logically different for the user, even if they then merge in the same interface. So It's important that they are immediately distinguishable. I've implemented the buttons "collapse" on small screen. Here the screenshots:
Large screen:
![myDspace-large](https://user-images.githubusercontent.com/20353276/86566220-945fd080-bf69-11ea-8bf8-f8e49e54745a.png)
Small screen:
![myDspace-small](https://user-images.githubusercontent.com/20353276/86566250-a2155600-bf69-11ea-9876-623f98dc89e8.png)

If you don't agree we propose to put it to a vote at the next meeting or to collect the preferences within this PR.
This emoticon for your solution :+1: , this for ours :100: .

I agree for the search button as primary (already fixed).

About the search input, I've splitted the buttons and the search input on small screen.
Large screen:
![external-large](https://user-images.githubusercontent.com/20353276/86566475-fd474880-bf69-11ea-91f9-448f78905d71.png)
Small screen:
![external-small](https://user-images.githubusercontent.com/20353276/86566484-033d2980-bf6a-11ea-8a08-7007ff4afe84.png)

The jumbotron is used as a placeholder (obviously it doesn't have to be a jumbotron): a block that tell the user where the results will be and that no search is performed yet. This is very useful because an empty space or a text like "No results" can be ambiguous; has a search already been done? Has there been an uncatched error? We can surely change the text to be more explicit: 'No search done yet'. Let me know.

About the rest api, that was agreed during the meeting, it is not possible to know in advance how the external source data will be used to create a workspaceitem and we agreed to keep the implementation for 7.0 as simple as possible. It was explicitely noted in the Tim comment in the jira issue and our implementation proposal.

The collection choice is now fixed on the PR https://github.com/DSpace/dspace-angular/pull/724. It was a cache problem (Commit https://github.com/4Science/dspace-angular/commit/6ef3f685caff72dcc61a8bddbdf366690fe6f5aa).

abollini commented:
I like the idea of a vote but it will be easier on a shorter comment and with default reaction.
Please 
- add +1 to this comment if you agree with our current implementation, i.e. two separate buttons to start a new submission or to import from an external source
- add -1 to this comment if you prefer to have a single dropdown button as in the @artlowel  comment https://github.com/DSpace/dspace-angular/pull/715#pullrequestreview-440003764

artlowel commented:
>I understand your concern but the two buttons should exist separately because the two processes are logically different for the user, even if they then merge in the same interface. So It's important that they are immediately distinguishable. 

I don't think the buttons as they are currently make it any clearer for the user. The fact that they're both next to the upload box gives the impression that they'll both start from an uploaded file. 

The first thing I thought when I saw the button was: If I upload a RIS or Bibtex file, I need to click the "import metadata" button to extract metadata form it. Without context, most people won't understand what it does without clicking it first. 

That's why the dropdown option in the mockup says "Import metadata from external source". Though I wouldn't recommend making the current button 3 times as wide just to fit that in.

I think a compromise could be: 
- Always use the icon-only buttons from your second screenshot. That would also make the buttons less obtrusive. 
- Give them tooltips on mouse over, that way we can get the text "Import metadata from external source" in there without issue. 
- And change the import icon to something that indicates the web, like a globe, to make it clear that this is not something that requires a file upload.


>The jumbotron is used as a placeholder (obviously it doesn't have to be a jumbotron): a block that tell the user where the results will be and that no search is performed yet. This is very useful because an empty space or a text like "No results" can be ambiguous; has a search already been done? Has there been an uncatched error? We can surely change the text to be more explicit: 'No search done yet'. Let me know.

I realize that the jumbotron was meant to be a placeholder, and I agree we shouldn't show an empty search result section.

But a big gray box with the text: "No search performed yet. Search items using external sources."  serves little purpose in my opinion. The only time you see it is when you just arrive on the page. So the user knows they haven't searched for anything.

If it said "Enter a query above to find items from the web to import in to DSpace" or something to that effect, it would be a little more useful as it would explain what the page is for. But personally, I'd just leave the page blank until there are search results.

>About the rest api, that was agreed during the meeting, it is not possible to know in advance how the external source data will be used to create a workspaceitem and we agreed to keep the implementation for 7.0 as simple as possible. It was explicitely noted in the Tim comment in the jira issue and our implementation proposal.

ok

>The collection choice is now fixed on the PR #724. It was a cache problem (Commit 4Science@6ef3f68).

👍

atarix83 commented:
@artlowel your feedback should be addressed, please could you review again ?

tdonohue commented:
Ugh, so this PR is also affected by merger of #541. :(  While it doesn't show merge conflicts here, this PR no longer builds properly against `main` and shows this error:
```
ERROR in src/app/submission/import-external/submission-import-external.component.ts:97:31 - error TS2339: Property 'SubmissionModal' does not exist on type 'typeof Context'.

97     this.context    = Context.SubmissionModal;
```

It looks like that `SubmissionModal` was for some reason removed in PR 541.  @artlowel : was this change accidental in 541?

atarix83 commented:
@tdonohue I fixed merge with main branch, now it should work, could you try? 
Currently the only external provider that seems to work is orcid, even if you try over official demo rest

atarix83 commented:
@tdonohue it seems a side effect of the problem described here https://github.com/DSpace/dspace-angular/issues/819

sourcedump commented:
> It's worth noting though that the REST API in #2824 works fine, so making ORCID REST requests like `http://localhost:8080/server/api/integration/externalsources/orcidV2/entries?query=Donohue` or even PubMed REST requests like `http://localhost:8080/server/api/integration/externalsources/pubmed/entries?query=mathematics` work perfectly OK. It's just that the UI doesn't seem to be making that same request?

Hello @tdonohue, I've made a new sync with the main branch to resolve conflicts and tested again the external import. I don't find any problem with the queries to the REST's endpoints. Ex.: for ORCID:
![orcid](https://user-images.githubusercontent.com/20353276/89873155-4b2f1c00-dbba-11ea-99b1-b10e6ec474eb.png)

and for Pubmed:
![pubmed](https://user-images.githubusercontent.com/20353276/89873200-5da95580-dbba-11ea-94aa-a66bc0e0be8a.png)

As you can see the requests are well formed.

tdonohue commented:
@sourcedump : It looks like recent changes to this PR have *decreased* the Code Coverage by over 52%?  I'm not sure why that would be, as it almost sounds like it's now skipping a large number of tests.  

Any ideas from you or maybe @artlowel  perhaps?  We cannot merge this until we determine how to correct the code coverage.

tdonohue commented on src/app/+my-dspace-page/my-dspace-new-submission/my-dspace-new-submission.component.spec.ts:
```diff
@@ -22,9 +22,11 @@ import { getMockScrollToService } from '../../shared/mocks/scroll-to-service.moc
 import { UploaderComponent } from 'src/app/shared/uploader/uploader.component';
 
-describe('MyDSpaceNewSubmissionComponent test', () => {
+fdescribe('MyDSpaceNewSubmissionComponent test', () => {
```
@sourcedump : The large code coverage drop is caused by this single line.  The `fdescribe` here adds a focus to this specific test & ensures that **only focused tests run**.  See https://jasmine.github.io/2.2/focused_specs.html

atarix83 replied:
fixed

tdonohue commented:
**NOTE FOR REVIEWERS:** As noted in today's meeting, this PR is actually *passing* Travis CI tests (even though it displays a red ❌).  Travis attempted to run its tests twice & the successful tests can be seen at https://github.com/DSpace/dspace-angular/pull/715/checks?check_run_id=1032397599

tdonohue commented:
@atarix83 : Yes, I agree with you. Let's resolve that extra `query` parameter on the REST side.  I'll create a ticket for it.  So, we can move forward this PR as-is.  Thanks for digging into the problem!

tdonohue commented:
Merging with 3 approvals.  (NOTE: The red ❌ on this PR is not accurate.  Travis strangely built this PR twice...it succeeded once, and then the second time GitHub thinks there was an error, even though clicking through the links shows it also succeeded).
